dali/operators/image/paste/multipaste.cc

```diff
@@ -0,0 +1,163 @@
+  auto &output = ws.template OutputRef<CPUBackend>(0);
+
+  output.SetLayout(images.GetLayout());
+  auto out_shape = output.shape();
+
+  auto& tp = ws.GetThreadPool();
+
+  auto batch_size = output.shape().num_samples();
+
+  TYPE_SWITCH(images.type().id(), type2id, InputType, (uint8_t, int16_t, int32_t, float), (
+      TYPE_SWITCH(output_type_, type2id, OutputType, (uint8_t, int16_t, int32_t, float), (
+          {
+            using Kernel = TheKernel<OutputType, InputType>;
+            for (int i = 0; i < batch_size; i++) {
+              auto paste_count = in_idx_[i].shape[0];
+
+              auto to_zero = view<OutputType, 3>(output[i]);
+              memset(to_zero.data, 0, to_zero.shape[0] * to_zero.shape[1] * to_zero.shape[2]);
+
+              if (no_intersections_[i]) {
+                for (int iter = 0; iter < paste_count; iter++) {
+                  int from_sample = in_idx_[i].data[iter];
+                  int to_sample = i;
+
+                  tp.AddWork(
+                      [&, i, iter, from_sample, to_sample](int thread_id) {
+                        kernels::KernelContext ctx;
+                        auto tvin = view<const InputType, 3>(images[from_sample]);
+                        auto tvout = view<OutputType, 3>(output[to_sample]);
+
+                        auto in_anchor_view = GetInAnchors(i, iter);
+                        auto in_shape_view = GetShape(i, iter, Coords(
+                            raw_input_size_mem_.data() + 2 * from_sample, dali::TensorShape<>(2)));
+                        auto out_anchor_view = GetOutAnchors(i, iter);
+
+                        kernel_manager_.Run<Kernel>(thread_id, to_sample, ctx, tvout, tvin,
+                                        in_anchor_view, in_shape_view, out_anchor_view);
+                      },
+                      out_shape.tensor_size(to_sample));
+                }
+              } else {
+                tp.AddWork(
+                    [&, i, paste_count](int thread_id) {
+                      for (int iter = 0; iter < paste_count; iter++) {
+                        int from_sample = in_idx_[i].data[iter];
+                        int to_sample = i;
+
+                        kernels::KernelContext ctx;
+                        auto tvin = view<const InputType, 3>(images[from_sample]);
+                        auto tvout = view<OutputType, 3>(output[to_sample]);
+
+                        auto in_anchor_view = GetInAnchors(i, iter);
+                        auto in_shape_view = GetShape(i, iter, Coords(
+                            raw_input_size_mem_.data() + 2 * from_sample, dali::TensorShape<>(2)));
+                        auto out_anchor_view = GetOutAnchors(i, iter);
+
+                        kernel_manager_.Run<Kernel>(thread_id, to_sample, ctx, tvout, tvin,
+                                        in_anchor_view, in_shape_view, out_anchor_view);
+                      }
+                    },
+                    paste_count);
```

Comment:
I'd unify the metric you use for execution priority. In line 132 you are using `out_shape.tensor_size(to_sample)` and here `paste_count`. This number represents the "amount of work" so that the bigger tasks are scheduled first to maximize thread utilization. The problem now is that you are comparing the tensor size which is a big number with paste_count, which is a small number, so you'd end up with the multi-paste case always being executed last.

You could change it to:
```
paste_count * out_shape.tensor_size(to_sample)
```
To represent that this is `paste_count` times more work than the one above.